accepted/future-releases/class-modifiers/feature-specification.md

```diff
@@ -530,62 +577,282 @@ mixinModifier     ::= 'sealed' | 'base' | 'interface' | 'final'
+does not come into play. Typedefs cannot be marked with any of the new
+modifiers.*
+
+### Disallowing implementation
+
+It is a compile-time error if a subtype of a declaration is marked `base` or
+`final` is not marked `base`, `final`, or `sealed`. This restriction applies to
+both direct and indirect subtypes and along all paths that introduce subtypes:
+`implements` clauses, `extends` clauses, `with` clauses, and `on` clauses. This
+restriction applies even to types within the same library.
+
+*Once the ability to use as an interface is removed, it cannot be reintroduced
+in a subtype. If a class is marked `base` or `final`, you may still implement
+the class's interface inside the same library, but the implementing class must
+again be marked `base`, `final`, or `sealed` to avoid it exposing an
+implementable interface.*
+
+Further, while you can ignore some restrictions on declarations within the same
+library, you can't use that to ignore restrictions inherited from other
+libraries.
+
+We say a class or mixin declaration `D` *can't be implemented locally* if it
+extends, mixes in, implements, or has as on type any declaration `S` where:
+
+*   `S` is from another library than `D`, and `S` has the modifier `base`,
+    `final` or `sealed`, or
+
+    ```dart
+    // a.dart
+    base mixin class S {}
+
+    // b.dart
+    import 'a.dart';
+
+    // These can't be implemented locally:
+    class DE extends S {}
+    class DM with S {}
+    mixin MO on S {}
+    ```
 
-    A _trivial generative constructor_ is a non-redirecting generative constructor 
-    which has
+*   `S` is from the same library as `D`, and `S` can't be implemented locally.
 
-    *   an empty parameter list,
-    *   no initializer list (no `: ...`),  
-    *   no constructor body (only `;`),
-    *   and is not marker `external`.
+    ```dart
+    // a.dart
+    base class B {}
+
+    // b.dart
+    import 'a.dart';
+
+    // This can't be implemented locally (from the previous rule):
+    base mixin class S extends B {}
+
+    // And thus these also can't be implemented locally (from this rule):
+    base class DE extends S {}
+    base class DM with S {}
+    base class DI implements S {}
+    base mixin MO on S {}
+    ```
+
+Otherwise, `D` can be implemented locally. It is a compile-time error if:
+
+*   A class or mixin declaration `D` can't be implemented locally, and `D` is
+    not marked `base`, `final` or `sealed`.
+
+    ```dart
+    // a.dart
+    base class B {}
+
+    // b.dart
+    import 'a.dart';
+
+    class D extends B {} // Error, needs `base`, `final`, or `sealed`.
+    ```
+
+    *Also:*
+
+    ```dart
+    // a.dart
+    base class B {}
+
+    // b.dart
+    import 'a.dart';
+
+    base class S extends B {} // Can't be implemented locally but OK.
+
+    class D extends S {} // Error, needs `base`, `final`, or `sealed`.
+    ```
+
+*   A class or mixin declaration `D` implements the interface of a class or
```

Comment:
By "implements the interface" here, you mean literally "with an `implements` clause" and not just "is a subtype of" right? This error is about inducing a subtype relation that does *not* inherit. Any subtype that does inherit (`extends`, `with`, or `on`) is OK?

Reply:
That is what I mean. I think that in the original I had something early saying "yadday implement (has the type in the `implements` clause) yadda ...".

However, I don't think it's consistent with how the language specification uses `implements X`, which means "have X as a superinterface", which can be both direct or indirect/transitive.

The spec says, about direct super-interfaces,
> ```latex
> A class has a set of \Index{direct superinterfaces}.
> This set contains the interface of its superclass
> and the interfaces of the classes specified in
> the \IMPLEMENTS{} clause of the class.
> ```

So, to be consistent, we should change 
> ... implements the interface of a class or mixin declaration `S`, declared in ...

to 

> ... specifies a class or mixin `S` as an element in the type list of its `implements` clause", `S` is declared in ...

@eernstg How to phrase?

Reply:
I really should have studied this PR carefully, but I did not manage to find the time.

Anyway, '`T` implements `S`' in spec-ese means that `T` is `T0` bounded for some `T0` (e.g., `X extends int` is `int` bounded, `int` is `int` bounded), and there is a superinterface path from `T0` to to `S` (not just a subtype relation). So `X extends List<int>` implements `Iterable<int>`, but not `Iterable<num>`.

I think the '`S` implements `T`' terminology works fine in this case, because we're generally talking about relationships that are created using clauses like `extends` and `implements`, and they're all covered.

If we want to specify that a given type `S` occurs in an `extends`, `with`, `implements`, or `on` clause in the declaration of `D` then we would simply say that `S` is a direct superinterface of `D`.